### PR TITLE
PostgreSQL: add initial support (WIP)

### PR DESCRIPTION
**Pull request by jufajardini**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/4241

Last PR (not exactly a continuation, though): https://github.com/OISF/suricata/pull/5938

Describe changes:
- Reboot the code: restart from scratch, using again setup-app-layer, to get rid of messy spaghetti code;
- Patch common files;
- rust/src/applayerpgsql: add nom parsers for decoding SSL Requests and Responses;
--- there's a TODO indicating a likely source of problem in how it parses SSL Responses, because other checks are necessary to make sure it is indeed an SSL Response instead of a longer, regular message.
- use a data model that allows adding more message types more easily
- Add nom parser for real startup messages.
- Adapt pgsql.rs to work with current pgsql nom parsers.
- This is parser parses SSL Requests, SSL Responses, and basic Startup Messages (parameter order still fixed) 

#suricata-verify-pr:
#suricata-verify-repo:
#suricata-verify-branch: https://github.com/jufajardini/suricata-verify/tree/ssl-handshake-startup-message-v0
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

File: rust/src/applayerpgsql/parser.rs [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+struct PgsqlParameter {
+    param_name: Vec<u8>,
+    param_value: Vec<u8>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PgsqlStartupParameters {
+    user: PgsqlParameter,
+    database: Option<PgsqlParameter>,
+    // TODO size-wise, being Option<Vec> or just Vec here is the same. I wonder if
```
Comment (victorjulien) [inline rust/src/applayerpgsql/parser.rs:36]: how did you reach this conclusion?

File: rust/src/applayerpgsql/pgsql.rs [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+    let eof = unsafe {
+        if AppLayerParserStateIssetFlag(pstate, APP_LAYER_PARSER_EOF_TS) > 0 {
+            true
+        } else {
+            false
+        }
+    };
+
+    if eof {
+        // If needed, handled EOF, or pass it into the parser.
```
Comment (victorjulien) [inline rust/src/applayerpgsql/pgsql.rs:360]: with eof we can still have data I think, so we'd have to process that

File: src/Makefile.am [hunk 3]
```diff
@@ -40,6 +40,7 @@ noinst_HEADERS = \
 	app-layer-nfs-udp.h \
 	app-layer-ntp.h \
 	app-layer-parser.h \
+    app-layer-pgsql.h \
```
Comment (victorjulien) [inline src/Makefile.am:43]: formatting looks off

**Comment by victorjulien** [comment]
I think adding more tests is the critical next step, as it will help you define the parsers and their data structures.

**Comment by victorjulien** [comment]
I'd also like these PRs to already fully pass our CI checks. In general leaving warnings in dev code is something I don't like. In many cases it has given me clues about things being broken than I missed because they got lost in the noise.

File: src/Makefile.am (hunk 3 above)
Comment (jufajardini) [reply src/Makefile.am:43]: Thank you, and I'm sorry for having to have that one pointed out >_<' . In VSCode they looked the same...

File: rust/src/applayerpgsql/parser.rs (hunk 1 above)
Comment (jufajardini) [reply rust/src/applayerpgsql/parser.rs:36]: As I was concerned about this, I searched and found an answer in StackOverflow that shared a macro where one could make such comparisons (here: https://stackoverflow.com/questions/16504643/what-is-the-overhead-of-rusts-option-type)
I used that macro to compare some possibilities, and that seemed to be the case, to me. Here are some of the comparisons I've tried: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=076232646a400fa0fe16a2fe0d53917a

(longer story here: https://jufajardini.wordpress.com/2021/03/10/til-sizes-of-structs-with-option-fields-vs-normal-attributes/)

Comment (jasonish) [reply rust/src/applayerpgsql/parser.rs:36]: `Vec<u8>` and `Option<Vec<u8>` do show the same size with `size_of`, but I think there is a piece missing.  `Vec<u8>` requires a `Vec::new()` which is going to use some more memory that doesn't appear to show up in `size_of`.

I wonder if a good rule of thumb for us would be:
- If under normal conditions, something will always be in the Vec, just initialize it.
- If its common enough that the list could be empty, wrap it in Option and allocate it as needed?

Comment (jufajardini) [reply rust/src/applayerpgsql/parser.rs:36]: Thanks for the clarification. I think that approach makes sense!
I was thinking yesterday that I could actually have something much simpler, like just one Vec where all parameters will be stored, after parsing, but not sure if that's good, and had some trouble in making that work with my current parsers. But if you'd consider that better, I'd find out how other implementations have done it, and make it work.

Comment (jufajardini) [reply rust/src/applayerpgsql/parser.rs:36]: In fact, considering memory allocation, having just one vec seems to make much more sense in my head, but as I'm still learning, I'll wait for more knowledgeable considerations x)

Comment (victorjulien) [reply rust/src/applayerpgsql/parser.rs:36]: While this is an interesting discussion, I would like to suggest to not worry about such things too much in this phase of the project. These kind of micro optimizations can easily add unnecessary complexity.

Comment (jufajardini) [reply rust/src/applayerpgsql/parser.rs:36]: Fair, and important advice, to someone like me :sweat_smile: I'll stay on the path with adding more tests, as you guided me to.

**Comment by jufajardini** [comment]
Closed with PR https://github.com/OISF/suricata/pull/6012

**Pull request closed without merging**